TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
-      auto theBranchName (branchName);
+   template <typename T = Double_t>
+   TActionResultPtr<Double_t> Mean(const std::string &branchName = "")
+   {
+      auto theBranchName(branchName);
       GetDefaultBranchName(theBranchName, "calculate the mean");
       auto meanV = std::make_shared<T>(0);
-      return CreateAction<T, EActionType::kMean>(theBranchName,meanV);
+      return CreateAction<T, Internal::EActionType::kMean>(theBranchName, meanV);
    }
 
 private:
+   /// Get the TDataFrameImpl if reachable. If not, throw.
+   std::weak_ptr<Details::TDataFrameImpl> GetDataFrameChecked()
+   {
+      auto df = fProxiedPtr->GetDataFrame();
+      if (df.expired()) {
+         throw std::runtime_error("The main TDataFrame is not reachable: did it go out of scope?");
+      }
+      return df;
+   }
 
-   void GetDefaultBranchName(std::string& theBranchName, const std::string& actionNameForErr) {
+   void GetDefaultBranchName(std::string &theBranchName, const std::string &actionNameForErr)
+   {
       if (theBranchName.empty()) {
          // Try the default branch if possible
-         const BranchList& defBl = fDerivedPtr->GetDataFrame().GetDefaultBranches();
+         const BranchList &defBl = fProxiedPtr->GetDataFrame().lock()->GetDefaultBranches();
          if (defBl.size() == 1) {
             theBranchName = defBl[0];
          } else {
-            std::string msg ("No branch in input to ");
+            std::string msg("No branch in input to ");
             msg += actionNameForErr;
-            msg +=  " and more than one default branch.";
+            msg += " and more than one default branch.";
             throw std::runtime_error(msg);
          }
       }
    }
 
-   template<typename BranchType, typename ActionResultType, enum EActionType, typename ThisType>
-   struct SimpleAction{ };
-
-   template<typename BranchType, typename ThisType>
-   struct SimpleAction<BranchType, TH1F, EActionType::kHisto1D, ThisType> {
-      static TActionResultPtr<TH1F> BuildAndBook(ThisType thisFrame, const std::string& theBranchName, std::shared_ptr<TH1F> h, unsigned nSlots) {
-         auto fillOp = std::make_shared<Operations::FillOperation>(h.get(), nSlots);
-         auto fillLambda = [fillOp](unsigned slot, const BranchType& v) mutable { fillOp->Exec(v, slot); };
+   template <typename BranchType, typename ActionResultType, enum Internal::EActionType, typename ThisType>
+   struct SimpleAction {};
+
+   template <typename BranchType, typename ThisType>
+   struct SimpleAction<BranchType, TH1F, Internal::EActionType::kHisto1D, ThisType> {
+      static TActionResultPtr<TH1F> BuildAndBook(ThisType thisFrame, const std::string &theBranchName,
+                                                 std::shared_ptr<TH1F> h, unsigned int nSlots)
+      {
+         // Here we pass the shared_ptr and not the pointer
+         auto fillOp = std::make_shared<Internal::Operations::FillOperation>(h, nSlots);
```

Comment:
we should comment on the fact that we need to pass a shared pointer to the operation to let the operation have the same scope of the lambda function (and therefore the action), so that the operation's destructor is called right after running the event loop.